### PR TITLE
documented execution flow

### LINKED ISSUE #737: RFC: document execution flow
When starting with mkosi it is hard to distinguish/order the actions that mkosi takes to construct an image,
especially since information on each step are separated over multiple locations in the man page
A simple flowchart would help to locate these steps and see in which context they are executed.

### PR DESCRIPTION
**Pull request by nopeslide**
fix #737 
It is hard to understand the mkosi execution flow,
since relevant information is scattered throughout the man page.
Therefore I added a simplified execution flow graph as ascii art to the man page.

### CODE REVIEW

**Comment by keszybz** [comment]
Thanks, this looks pretty nice.
The only issue is where in the man page this new section should go.
Right now it is between "Command line parameters" and "Command line parameters and their settings file counterparts", which seems like the worst possible place. Can you move after "Commands" instead?

(Why there: In general, an overview chart like this should be as early as possible, because the man page should start with general concepts and the "drill down" into details. But this chart only applies to `build` verb. So we can't show the chart before introducing `build`, which is in "Commands".)

**Comment by keszybz** [comment]
Oh, and maybe also add a sentence like:
This is the workflow when `build` is invoked.

File: mkosi.md [hunk 1]
```diff
@@ -937,6 +937,110 @@ details see the table below.
...
+      |                 .         packages         .
+      |                 .            |             .
+      |                 .            v             .
+      |                 .         install          .
+      |                 .      build packages      .
+      |                 .            |             .
+      |                 .            v             .
+      |                 .           run            .
+      |                 .      prepare script      .
+      |                 .         (build)          .
```
Comment (keszybz) [inline mkosi.md:967]: Maybe mention `mkosi.prepare` here? E.g. change
```diff
- (build)
+ (`mkosi.prepare build`)
```

File: mkosi.md [hunk 2]
```diff
@@ -937,6 +937,110 @@ details see the table below.
...
+      |                 .       extra trees        .
+      |                 .            |             .
+      |                 .            v             .
+      |                 .          copy            .
+      |                 .      build sources       .
+      |                 .            |             .
+      |                 .            v             .
+      |                 .           run            .
+      |                 .     postinstall script   .
+      |                 .         (build)          .
```
Comment (keszybz) [inline mkosi.md:980]: `(mkosi.postinstall build)` ?

File: mkosi.md [hunk 3]
```diff
@@ -937,6 +937,110 @@ details see the table below.
...
+      |        v        .                          .
+      |       run       .                          .
+      | finalize script .                          .
+      |     (build)     .                          .
+      |        |        .                          .
+      |        '---------------------.             .
+      |                 .            |             .
+      |                 .            v             .
+      |                 .           run            .
+      |                 .       build script       .
```
Comment (keszybz) [inline mkosi.md:993]: ...

File: mkosi.md [hunk 4]
```diff
@@ -937,6 +937,110 @@ details see the table below.
...
+                        .                          .       distribution
+                        .                          .            |
+                        .                          .            v
+                        .                          .         install
+                        .                          .         packages
+                        .                          .            |
+                        .                          .            v
+                        .                          .           run
+                        .                          .      prepare script
+                        .                          .         (final)
```
Comment (keszybz) [inline mkosi.md:1012]: ...

File: mkosi.md [hunk 5]
```diff
@@ -937,6 +937,110 @@ details see the table below.
...
+                        .                          .       extra trees
+                        .                          .            |
+                        .                          .            v
+                        .                          .          copy
+                        .                          .   build destination
+                        .                          .            |
+                        .                          .            v
+                        .                          .           run
+                        .                          .     postinstall script
+                        .                          .         (final)
```
Comment (keszybz) [inline mkosi.md:1025]: ...

File: mkosi.md [hunk 6]
```diff
@@ -937,6 +937,110 @@ details see the table below.
...
+                        .                          .            |
+               .------------------------------------------------'
+               |        .                          .
+               v        .                          .
+              run       .                          .
+        finalize script .                          .
+            (final)     .                          .
```
Comment (keszybz) [inline mkosi.md:1032]: ...

**Review by keszybz — COMMENTED** [review]
I think it'd be nice to add the (default) script names, to make it easy to figure out which file one wants to edit to change some phase of the build.

**Comment by behrmann** [comment]
Just ran `pandoc -t man -s -o man/mkosi.1 mkosi.md` on it to generate the man page and it also comes out nicely there. Adding a short paragraph before the flowchart does improve the placement within the section, though.

**Comment by behrmann** [comment]
This looks pretty great. I think you could still add the default path names for `mkosi.extra` and `mkosi.skeleton` and squash the commits, but apart from that, this is really nice!

**Comment by bcbnz** [comment]
I think it would be useful to add the points at which the cached images are stored/loaded for incremental builds (after the prepare script is run). Maybe it should also be noted (afterwards, it might be confusing to try and add it in the chart) that the first run of an incremental build constructs both cached images first and then continues the rest of the build.

**Comment by nopeslide** [comment]
> I think it would be useful to add the points at which the cached images are stored/loaded for incremental builds (after the prepare script is run). Maybe it should also be noted (afterwards, it might be confusing to try and add it in the chart) that the first run of an incremental build constructs both cached images first and then continues the rest of the build.

I left the incremental build out by intention, because it would be quite confusing because it breaks the "logical" flow.

**Comment by nopeslide** [comment]
> This looks pretty great. I think you could still add the default path names for `mkosi.extra` and `mkosi.skeleton` and squash the commits, but apart from that, this is really nice!

good point. I added the default paths

**Comment by bcbnz** [comment]
> I left the incremental build out intentionally.
> the result would be quite confusing because it breaks the "logical" flow.

OK. Could it be mentioned in the text before or after the diagram? I feel it is an important part of the execution flow and should be part of this section.

Nice diagram by the way, I am finding it quite useful as a source for documenting one of our repositories which builds images.

**Comment by nopeslide** [comment]
> OK. Could it be mentioned in the text before or after the diagram? I feel it is an important part of the execution flow and should be part of this section.

Is this alright? I tried to solve it textual, so it doesn't complicate the diagram itself

**Comment by bcbnz** [comment]
I couldn't get a review on the commit to work properly, but how about

> When building with \`--incremental\` mkosi creates a cache of both images
> after running the prepare script on the first run, and starts from this point in
> subsequent runs.

instead? It is a bit more explicit about when it creates the cache and how it uses it.

**Comment by nopeslide** [comment]
> I couldn't get a review on the commit to work properly, but how about
> 
> > When building with `--incremental` mkosi creates a cache of both images
> > after running the prepare script on the first run, and starts from this point in
> > subsequent runs.
> 
> instead? It is a bit more explicit about when it creates the cache and how it uses it.

I skimmed the code and it seems mkosi skips every hook when building the cached image and only installs the distribution related stuff.
When using the cached image it copies data from the cached image and skips the distribution installation.
Did I misunderstand sth?

**Comment by nopeslide** [comment]
I commented out the lines that shouldn't be triggered in each situation by simply looking if `cached_tree` or `for_cache` is used as parameter

when building the cache:
```python
def build_image(...)
    ...
    cached_tree = reuse_cache_tree(args, root, do_run_build_script, for_cache, cached)
    install_skeleton_trees(args, root)
    install_distribution(args, root, do_run_build_script, cached_tree)
    install_etc_hostname(args, root, cached_tree)
    install_boot_loader(args, root, loopdev, do_run_build_script, cached_tree)
    run_prepare_script(args, root, do_run_build_script, cached_tree)
    #install_extra_trees(args, root, for_cache)
    #install_build_src(args, root, do_run_build_script, for_cache)
    #install_build_dest(args, root, do_run_build_script, for_cache)
    set_root_password(args, root, do_run_build_script, cached_tree)
    set_serial_terminal(args, root, do_run_build_script, cached_tree)
    set_autologin(args, root, do_run_build_script, cached_tree)
    sshkey = setup_ssh(args, root, do_run_build_script, for_cache, cached_tree)
    setup_network_veth(args, root, do_run_build_script, cached_tree)
    #run_postinst_script(args, root, loopdev, do_run_build_script, for_cache)

    if cleanup:
        clean_package_manager_metadata(root)
        remove_files(args, root)
    #reset_machine_id(args, root, do_run_build_script, for_cache)
    reset_random_seed(args, root)
    #run_finalize_script(args, root, do_run_build_script, for_cache)
    #invoke_fstrim(args, root, do_run_build_script, for_cache)
    #make_read_only(args, root, for_cache)
    ...
```
when using the cache
```python
def build_image(...)
    ...
    cached_tree = reuse_cache_tree(args, root, do_run_build_script, for_cache, cached)
    install_skeleton_trees(args, root)
    #install_distribution(args, root, do_run_build_script, cached_tree)
    #install_etc_hostname(args, root, cached_tree)
    #install_boot_loader(args, root, loopdev, do_run_build_script, cached_tree)
    #run_prepare_script(args, root, do_run_build_script, cached_tree)
    install_extra_trees(args, root, for_cache)
    install_build_src(args, root, do_run_build_script, for_cache)
    install_build_dest(args, root, do_run_build_script, for_cache)
    #set_root_password(args, root, do_run_build_script, cached_tree)
    #set_serial_terminal(args, root, do_run_build_script, cached_tree)
    set_autologin(args, root, do_run_build_script, cached_tree)
    #sshkey = setup_ssh(args, root, do_run_build_script, for_cache, cached_tree)
    #setup_network_veth(args, root, do_run_build_script, cached_tree)
    run_postinst_script(args, root, loopdev, do_run_build_script, for_cache)

    if cleanup:
        clean_package_manager_metadata(root)
        remove_files(args, root)
    reset_machine_id(args, root, do_run_build_script, for_cache)
    reset_random_seed(args, root)
    run_finalize_script(args, root, do_run_build_script, for_cache)
    invoke_fstrim(args, root, do_run_build_script, for_cache)
    make_read_only(args, root, for_cache)
    ...
```

**Comment by bcbnz** [comment]
It certainly runs the prepare script -- I just booted one of the cached images I have here and the changes made by prepare exist in the cached image. Also, the current manpage says 

> --prepare-script=
>           Takes  a  path  to an executable that is invoked inside the image right after installing the software packages.  It is
>            the last step before the image is cached (if incremental mode is enabled).

This is using raw image outputs, somebody more familiar with the code may be able to comment if there are differences with other image types.

**Comment by bcbnz** [comment]
And the skeleton files are in the cached image as well.

**Comment by nopeslide** [comment]
> And the skeleton files are in the cached image as well.

they are copied regardless if ached or not

**Comment by nopeslide** [comment]
> > --prepare-script=
> > Takes  a  path  to an executable that is invoked inside the image right after installing the software packages.  It is
> > the last step before the image is cached (if incremental mode is enabled).

good catch. I merged the steps

**Comment by bcbnz** [comment]
> > And the skeleton files are in the cached image as well.
> 
> they are copied regardless if cached or not

You're correct, I hadn't noticed. Output from one of my builds confirms that.

> ‣ Copied cached image as ...
> ‣ Attaching image file...
> ‣ Attached image file as /dev/loop0.
> ‣ Mounting image...
> ‣ Copying in skeleton file trees...
> ‣ Copying in extra file trees...
> ‣ Copying in build script...
> ‣ Copying in sources...
> ‣ Mounting Package Cache...
> ‣ Running postinstall script...

I would actually argue this is a bug -- if the package installation or the prepare script (which is not run again after loading the cache) changes any of the files from the skeleton tree (e.g., if the skeleton tree was an initial configuration for something which is then updated) then re-copying would overwrite those changes. I guess that should probably be the topic of a different discussion though.

**Comment by nopeslide** [comment]
@bcbnz more thoughts? would squash otherwise

**Review by keszybz — COMMENTED** [review]
Looks great, thanks.

**Pull request merged**